automation/deploy_igz/deploy_remote.py

```diff
@@ -0,0 +1,293 @@
+#
+
+import io
+import logging
+import os
+import re
+import subprocess
+import sys
+from typing import List
+
+import click
+import coloredlogs
+import paramiko
+import yaml
+
+log_level = logging.INFO
+fmt = "%(asctime)s %(levelname)s %(message)s"
+logging.basicConfig(level=log_level)
+logger = logging.getLogger("mlrun-deploy")
+coloredlogs.install(level=log_level, logger=logger, fmt=fmt)
+
+
+class MlRunDeployer(object):
+    class Consts(object):
+        mandatory_fields = ["DATA_NODES", "USER", "PASSWORD", "DOCKER_REGISTRY"]
+
+    def __init__(self, conf_file):
+        self._config = yaml.safe_load(conf_file)
+        for key in self.Consts.mandatory_fields:
+            if self._config.get(key, None) is None:
+                raise RuntimeError(f"Mandatory option {key} not defined")
+        self._username = self._config.get("REGISTRY_USERNAME")
+        self._password = self._config.get("REGISTRY_PASSWORD")
+        if self._username is not None and self._password is None:
+            raise RuntimeError(
+                "REGISTRY_USERNAME defined, yet REGISTRY_PASSWORD is not defined"
+            )
+
+    @staticmethod
+    def _get_image_tag(branch, tag):
+        return f"v{tag}-{branch}"
+
+    @staticmethod
+    def _execute_local_proc_interactive(cmd):
+        proc = subprocess.Popen(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True
+        )
+        for line in proc.stdout:
+            yield line
+        proc.stdout.close()
+        ret_code = proc.wait()
+        if ret_code:
+            raise subprocess.CalledProcessError(ret_code, cmd)
+
+    def _exec_local(self, cmd: List[str], live=False) -> str:
+        logger.debug("Exec local: %s", " ".join(cmd))
+        buf = io.StringIO()
+        for line in self._execute_local_proc_interactive(cmd):
+            buf.write(line)
+            if live:
+                print(line, end="")
+        output = buf.getvalue()
+        return output
+
+    def _exec_remote(self, cmd: List[str], live=False):
+        cmd_str = " ".join(cmd)
+
+        logger.debug("Exec remote: %s", cmd_str)
+
+        stdin_stream, stdout_stream, stderr_stream = self._ssh_client.exec_command(
+            cmd_str
+        )
+
+        stdout = ""
+        if live:
+            while True:
+                line = stdout_stream.readline()
+                stdout += line
+                if not line:
+                    break
+                print(line, end="")
+        else:
+            stdout = stdout_stream.read().decode("utf8")
+
+        stderr = stderr_stream.read().decode("utf8")
+
+        exit_status = stdout_stream.channel.recv_exit_status()
+
+        if exit_status:
+            raise RuntimeError(
+                f"Command '{cmd_str}' finished with failure ({exit_status})\n{stderr}"
+            )
+
+        return stdout
+
+    def _get_current_version(self) -> str:
+        sys.path.append(os.path.dirname(os.path.abspath(__file__)) + "/..")
```

Comment:
I have an idea for this which will make it more readable. You can create a context manager method:
```python
@contextlib.contextmanager
def add_mlrun_src_to_path(self):
    mlrun_src = os.path.dirname(os.path.abspath(__file__)) + "/.."
    sys.path.append(mlrun_src)
    yield sys.path
    sys.path.remove(mlrun_src)
```

Then you can:
```python
with self.add_mlrun_src_to_path():
    from version import version_file
    return str(version_file.read_unstable_version_prefix())
```